basictracer/tracer.py

```diff
@@ -87,7 +90,14 @@ def start_span(self,
                 else child_of.context)
         elif references is not None and len(references) > 0:
             # TODO only the first reference is currently used
-            parent_ctx = references[0].referenced_context
+            first_ref = references[0]
+            if not isinstance(first_ref, opentracing.Reference):
+                msg = (
+                    'references[0] should be a opentracing.Reference '
+                    'objects, got %r instead'
+                )
+                raise TypeError(msg % first_ref)
```

Comment:
I would not recommend raising an error from this method. Can log an error and ignore the reference

Reply:
@yurishkuro raising the error here is good, the current behavior causes a ValueError to be raised with a very cryptic message.

Reply:
To reproduce the issue:

```python
import opentracing
import uuid

parent = opentracing.tracer.extract(
    opentracing.Format.TEXT_MAP,
    {
        "ot-tracer-spanid": uuid.uuid4().hex,
        "ot-tracer-sampled": "true",
        "ot-tracer-traceid": uuid.uuid4().hex,
    },
)

child = opentracing.tracer.start_span(
    operation_name="start new span as child",
    references=opentracing.child_of(parent),
)
```

The typo is subtle here, but pernicious and time-consuming because the current exception raised does not explain the problem.

Raising a TypeError shows exactly where the problem is and engineers can fix easily like so:


```diff
child = opentracing.tracer.start_span(
    operation_name="start new span as child",
-    references=opentracing.child_of(parent),
+    references=[opentracing.child_of(parent)],
)
```

Reply:
Furthermore, this issue only arises because the implementation of the type `Reference` from *opentracing-python* [inherits from a `NamedTuple`](https://github.com/opentracing/opentracing-python/blob/946f763510a8c26a66d5c53a9c4d0c6027cb9129/opentracing/tracer.py#L253-L255)

The error is misleading because `references[0]` is a valid call since Reference behaves like a tuple.

I would suggest to change the `opentracing.Reference` implementation to be a its own type instead, preventing this issue.

I'm glad to open a PR to opentracing-python to implement immutability without inheriting from a NamedTuple [created on-the-fly](https://github.com/opentracing/opentracing-python/blob/946f763510a8c26a66d5c53a9c4d0c6027cb9129/opentracing/tracer.py#L255).

Perhaps something like along the lines of:

```python
class Reference(object):
    """A Reference pairs a reference type with a referenced :class:`SpanContext`.
    References are used by :meth:`Tracer.start_span()` to describe the
    relationships between :class:`Span`\\ s.
    :class:`Tracer` implementations must ignore references where
    referenced_context is ``None``.  This behavior allows for simpler code when
    an inbound RPC request contains no tracing information and as a result
    :meth:`Tracer.extract()` returns ``None``::
        parent_ref = tracer.extract(opentracing.HTTP_HEADERS, request.headers)
        span = tracer.start_span(
            'operation', references=child_of(parent_ref)
        )
    See :meth:`child_of` and :meth:`follows_from` helpers for creating these
    references.
    """
    def __init__(self, type, referenced_context):
        self.type = type
        self.referenced_context = referenced_context

    def __getitem__(self, position):
        """backwards-compatibility with NamedTuple implementation"""
        if position == 0:
            return self.type
        if position == 1:
            return self.referenced_context

        raise IndexError("Reference supports only 2 positions: 0 (type) or 1 (referenced_context)")
```

Reply:
Would it be helpful ?